src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/editor/Workspace.kt

```diff
@@ -24,44 +29,78 @@ import org.evosuite.utils.CompactReport
 class Workspace(private val project: Project) {
     data class TestJobInfo(val fileUrl: String, var targetUnit: String, val modificationTS: Long, val jobId: String)
 
+    private class TestJob(val info: TestJobInfo, val report: CompactReport, val selectedTests: HashSet<String>) {
+        fun getSelectedTests(): List<CompactTestCase> {
+            return report.testCaseList.filter { selectedTests.contains(it.key) }.map { it.value }
+        }
+
+        fun getSelectedLines(): HashSet<Int> {
+            val lineSet: HashSet<Int> = HashSet()
+            getSelectedTests().map { lineSet.addAll(it.coveredLines) }
+            return lineSet
+        }
+    }
+
     private val log = Logger.getInstance(this.javaClass)
 
     /**
      * Maps a workspace file to the test generation jobs that were triggered on it.
      * Currently, the file key is represented by its presentableUrl
      */
-    private val testGenerationResults: HashMap<String, ArrayList<Pair<TestJobInfo, CompactReport>>> = HashMap()
+    private val testGenerationResults: HashMap<String, ArrayList<TestJob>> = HashMap()
 
     /**
      * Maps a test generation job id to its corresponding test job information
      */
     private var pendingTestResults: HashMap<String, TestJobInfo> = HashMap()
 
     init {
+        val connection = project.messageBus.connect()
+
+        // Set event listener for changes to the VFS. The overridden event is
+        // triggered whenever the user switches their editor window selection inside the IDE
+        connection.subscribe(
+            COVERAGE_SELECTION_TOGGLE_TOPIC,
+            object : CoverageSelectionToggleListener {
+                override fun testGenerationResult(testName: String, selected: Boolean, editor: Editor) {
+                    val vFile = vFileForDocument(editor.document) ?: return
+                    val fileKey = vFile.presentableUrl
+                    val testJob = testGenerationResults[fileKey]?.last() ?: return
+                    val modTs = editor.document.modificationStamp
+
+                    if (selected) {
+                        testJob.selectedTests.add(testName)
+                    } else {
+                        testJob.selectedTests.remove(testName)
+                    }
+
+                    // update coverage only if the modification timestamp is the same
+                    if (testJob.info.modificationTS == modTs) {
```

Comment:
Why do we need to check for `modificationTS`?

Reply:
so that we update coverage only if the modification timestamp is the same